core/extraction/round_two.py

```diff
@@ -806,29 +811,109 @@ def extract_outcomes(df_input: pd.DataFrame) -> pd.DataFrame:
+
+    # Get data for each financial half as its own row
+    df_unpivoted_financial_years = pd.melt(
+        unpivoted_df,
+        id_vars=id_vars,
+        value_vars=financial_years,
+        value_name="Amount",
+        var_name="financial_period",
+    )
+
+    df_unpivoted_financial_years["financial_year"] = df_unpivoted_financial_years["financial_period"].str.extract(
+        r"(\d{2}/\d{2})"
+    )
+
+    financial_year_start = pd.to_datetime(
+        "20" + df_unpivoted_financial_years["financial_year"].str[:2] + "-04-01", format="%Y-%m-%d"
+    )
+    financial_year_end = financial_year_start + pd.DateOffset(years=1) - pd.DateOffset(days=1)
+
+    # Create an empty list to store the new rows
+    new_rows = []
+
+    # Iterate over each row in the original dataframe
+    for index, row in df_unpivoted_financial_years.iterrows():
+        start_date = financial_year_start[index]
+        end_date = financial_year_end[index]
+
+        new_rows.append(
+            {
+                "start_date": start_date,
+                "end_date": end_date,
+            }
         )
-        # up to 10 projects per row
-        for idx in range(0, 10):
-            # 15 cols per project "section"
-            col_idx = idx * 15
-            proj_outcome = pd.DataFrame().append(flat_row.iloc[col_idx + 3 : col_idx + 18])
 
-            # skip iteration if project field is empty
-            if not proj_outcome.iloc[:, 2].any():
-                continue
-            proj_outcome.columns = col_headers
-            proj_outcome = pd.concat([temp_programme, proj_outcome], axis=1)
-            prog_proj_outcomes = prog_proj_outcomes.append(proj_outcome, ignore_index=True)
+    temp_df = pd.DataFrame(new_rows)
 
-        df_outcomes_out = df_outcomes_out.append(prog_proj_outcomes)
+    df_unpivoted_financial_years["Start_Date"] = temp_df["start_date"]
+    df_unpivoted_financial_years["End_Date"] = temp_df["end_date"]
 
-    return df_outcomes_out
+    final_cols = [
+        "Submission ID",
+        "Project ID",
+        "Programme ID",
+        "Outcome",
+        "Start_Date",
+        "End_Date",
+        "UnitofMeasurement",
+        "GeographyIndicator",
+        "Amount",
+        "Actual/Forecast",
+        "Higher Frequency",
+    ]
+
+    df_outcomes = df_unpivoted_financial_years.rename(
+        columns={
+            "Tab 6 - Outcomes: Section B - Geography": "GeographyIndicator",
+            "Tab 6 - Outcomes: Section B - Indicator Name": "Outcome",
+            "Tab 6 - Outcomes: Section B - Unit of Measurement": "UnitofMeasurement",
+            "Tab 6 - Outcomes: Section B - Higher Frequency": "Higher Frequency",
+        }
+    )
+
+    df_outcomes["Actual/Forecast"] = df_outcomes.apply(get_actual_forecast, axis=1)
+    df_outcomes["Higher Frequency"] = df_outcomes["Higher Frequency"].replace(0.0, np.nan)
+
+    df_outcomes = df_outcomes[final_cols]
+
+    return df_outcomes
```

Comment:
as a general observation about this function, there is a lot of local variables, some of which are quite large. We probably get away with it here as it's a rarely used data pipeline, but, for instance, this function has 8 large dataframes in local namespace by the time it gets to the end, and this is as a general rule pretty bad practice (apologies if some of this was inherited from the mess I left btw). But for some of them, there is no actual need to create a new dataframe, for example 

unpivoted_df -> df_unpivoted_financial_years -> df_outcomes

Each time these new df's are created, it is effectively creating an extra copy that it holds in memory, but the one before is never used again, so you could just reuse the variable name each time.  Does that make sense, happy to elaborate more if not?